element_array_ephys/ephys.py

```diff
@@ -556,75 +581,93 @@ def make(self, key):
+                            'waveform_mean': channel_waveform.mean(axis=0),
+                            'waveforms': channel_waveform})
+                        if channel2electrodes[channel]['electrode'] == unit_dict['electrode']:
+                            unit_peak_waveform = {
+                                **unit_dict,
+                                'peak_chn_waveform_mean': channel_waveform.mean(axis=0)}
+
+                    yield unit_peak_waveform, unit_electrode_waveforms
+
+        # insert waveform on a per-unit basis to mitigate potential memory issue
+        self.insert1(key)
+        for unit_peak_waveform, unit_electrode_waveforms in yield_unit_waveforms():
+            self.Unit.insert1(unit_peak_waveform, ignore_extra_fields=True)
+            self.UnitElectrode.insert(unit_electrode_waveforms, ignore_extra_fields=True)
 
 
 # ----------- Quality Control ----------
 
+
 @schema
-class ClusterQualityMetrics(dj.Imported):
-    definition = """
-    -> CuratedClustering.Unit
-    ---
-    amp: float
-    snr: float
-    isi_violation: float
-    firing_rate: float
-
-    presence_ratio: float  # Fraction of epoch in which spikes are present
-    amplitude_cutoff: float  # Estimate of miss rate based on amplitude histogram
-    isolation_distance=null: float  # Distance to nearest cluster in Mahalanobis space
-    l_ratio=null: float  # 
-    d_prime=null: float  # Classification accuracy based on LDA
-    nn_hit_rate=null: float  # 
-    nn_miss_rate=null: float
-    silhouette_score=null: float  # Standard metric for cluster overlap
-    max_drift=null: float  # Maximum change in spike depth throughout recording
-    cumulative_drift=null: float  # Cumulative change in spike depth throughout recording 
+class QualityControl(dj.Imported):
+    definition = """  # Quality control metrics for a set of CuratedClustering
+    -> CuratedClustering
     """
 
-    @property
-    def key_source(self):
-        return Clustering
+    class ClusterQualityMetrics(dj.Part):
+        definition = """  # Quality metrics for each cluster
+        -> master
+        -> CuratedClustering.Unit
+        ---
+        amp: float
+        snr: float
+        isi_violation: float
+        firing_rate: float
+        presence_ratio: float  # Fraction of epoch in which spikes are present
+        amplitude_cutoff: float  # Estimate of miss rate based on amplitude histogram
+        isolation_distance=null: float  # Distance to nearest cluster in Mahalanobis space
+        l_ratio=null: float  # 
+        d_prime=null: float  # Classification accuracy based on LDA
+        nn_hit_rate=null: float  # 
+        nn_miss_rate=null: float
+        silhouette_score=null: float  # Standard metric for cluster overlap
+        max_drift=null: float  # Maximum change in spike depth throughout recording
+        cumulative_drift=null: float  # Cumulative change in spike depth throughout recording 
+        """
```

Comment:
Are these quality metrics specific for ks2? To make it more general, we may consider doing something like:
```
class UnitQualityMetrics(dj.Part):
definition = """
-> master
-> CuratedClustering.Unit
"""
class UnitQualityMetric(dj.Part):
definition = """
-> master.UnitQualityMetrics
metric_name:  varchar(32)
---
metric_value=null: float
"""
If these metrics are universal, it's better to spell them out.